### PR TITLE
llvm: upstream Linux fixes

### PR DESCRIPTION
**Pull request by danielnachun**
- [ ] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [ ] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [ ] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

File: Formula/llvm.rb
Comment (carlocab) [inline Formula/llvm.rb:50]: Is it worth adding DSL for this in `brew`, to simplify writing it repeatedly across different formulae?

Comment (iMichka) [reply Formula/llvm.rb:50]: I think we could add one. Like the macos sdk DSLs. We can merge this PR and work on the DSL when we have time for it.

File: Formula/llvm.rb
Comment (carlocab) [inline Formula/llvm.rb:178]: Their build scripts generate an `install/strip` target. Wouldn't it be better to call that target instead on Linux? Or maybe calling it on both, though I'd want to know what this changes on macOS.

File: Formula/llvm.rb
Comment (danielnachun) [reply Formula/llvm.rb:50]: Agreed on this, that `if` statement is quite complicated and should be simplified.  I'm not actually sure that second check is necessary - if your `glibc` version is too old, that is the first formula which gets installed when you set up Homebrew for the first time.

Comment (carlocab) [reply Formula/llvm.rb:50]: > if your `glibc` version is too old, that is the first formula which gets installed when you set up Homebrew for the first time.

In this case, we should probably drop the second check. The only time that the second check is needed is when `glibc` isn't getting installed when it needed to be. However, whenever that's true, we should be fixing things in `brew` and not papering over them with workarounds in formulae.

**Comment by danielnachun** [comment]
Any thoughts on the failure on Catalina?

```
/usr/local/Cellar/cmake/3.20.4/bin/cmake -E create_symlink /tmp/llvm-20210617-34972-1r49mjd/llvm-project-12.0.0.src/compiler-rt/lib/builtins/aarch64/lse.S /tmp/llvm-20210617-34972-1r49mjd/llvm-project-12.0.0.src/llvm/build/runtimes/builtins-bins/outline_atomic_helpers.dir/outline_atomic_swp2_3.S
15869
CMake Error: failed to create symbolic link '/tmp/llvm-20210617-34972-1r49mjd/llvm-project-12.0.0.src/llvm/build/runtimes/builtins-bins/outline_atomic_helpers.dir/outline_atomic_swp2_3.S': file already exists
15870
make[5]: *** [outline_atomic_helpers.dir/outline_atomic_swp2_3.S] Error 1
15871
make[5]: *** Deleting file `outline_atomic_helpers.dir/outline_atomic_swp2_3.S'
15872
make[4]: *** [CMakeFiles/clang_rt.cc_kext_arm64_osx.dir/all] Error 2
15873
make[4]: *** Waiting for unfinished jobs....
```

This isn't something that should be affected by the changes I made, so I'm not sure why this wasn't a problem before.

File: Formula/llvm.rb
Comment (danielnachun) [reply Formula/llvm.rb:50]: Yes, my thoughts exactly.  That check goes back to `llvm@7` and it may be irrelevant now.

**Comment by carlocab** [comment]
> Any thoughts on the failure on Catalina?

Just run it again; the build is a bit temperamental...

File: Formula/llvm.rb
Comment (carlocab) [reply Formula/llvm.rb:50]: Without the second check this could just be
```suggestion
    depends_on "glibc" if Formula["glibc"].any_version_installed?
```
and we may not even need any new DSL for this. We would also no longer need the
```ruby
require "os/linux/glibc"
```

Comment (danielnachun) [reply Formula/llvm.rb:50]: I think I'm going to clean both the `require` statement and second check out of all versions of the formula in that case.

Comment (carlocab) [reply Formula/llvm.rb:50]: I'm actually also wondering why the `glibc` dependency is handled in superenv whenever it's installed, rather than having to do
```ruby
depends_on "glibc" if Formula["glibc"].any_version_installed?
```
in various formulae. That can probably be addressed separately, however.

Comment (danielnachun) [reply Formula/llvm.rb]: I cleaned all of this up now (apologies again for backing up the CI queue - I should have spaced out my updates for this).  

@iMichka do you want me to clean this up in linuxbrew/core before we merge here?  Or will the changes be obvious enough based to fix when you merge from homebrew/core to linuxbrew/core.

File: Formula/llvm.rb
Comment (carlocab) [inline Formula/llvm.rb:393]: This patch is a bit big. Can we put this in Homebrew/formula-patches?

File: Formula/llvm.rb
Comment (carlocab) [inline Formula/llvm.rb:286]:
```suggestion
                    (opt_lib/shared_library("libc++", "1")).to_s
```

File: Formula/llvm.rb
Comment (carlocab) [inline Formula/llvm.rb:130]: If you set this note that CMake will ignore the `CFLAGS` and `CXXFLAGS` environment variables.

**Comment by danielnachun** [comment]
I've pushed the latest changes we've made back in linuxbrew/core into this now.  Here's what's been updated:

1) We were able to build LLVM with GCC 5, which simplifies the dependency tree of formulae which which depend on LLVM.  I did have to attach a patch to the file, but as noted in the comment, this was provided by upstream and will be added to the next point release, after which it can be removed.
2) I used the `install/strip` target.  This update will serve as a test of using this on macOS, as I didn't make it Linux-only.  There were no issues I could see when using this on Linux, and hopefully the same will be true on macOS.

Things have cleaned up very nicely now for this PR so I think it should be ready soon!

File: Formula/llvm.rb
Comment (danielnachun) [reply Formula/llvm.rb:393]: Sure, I'll move it there.

File: Formula/llvm.rb
Comment (danielnachun) [reply Formula/llvm.rb:130]: Ah, so it's probably better to use `ENV.append` for this instead.

File: Formula/llvm.rb
Comment (carlocab) [inline Formula/llvm.rb:50]:
```suggestion
    patch do
      url "https://raw.githubusercontent.com/Homebrew/formula-patches/f0b8ff8b7ad4c2e1d474b214cd615a98e0caa796/llvm/llvm.patch"
      sha256 "084adce7711b07d94197a75fb2162b253186b38d612996eeb6e2bc9ce5b1e6e2"
    end
```

File: Formula/llvm.rb
Comment (carlocab) [inline Formula/llvm.rb:489]:
```suggestion
```

File: Formula/llvm.rb
Comment (carlocab) [inline Formula/llvm.rb:164]: Might be good to use `install-xcode-toolchain/strip` here too. If there is such a target, that is.

File: Formula/llvm.rb
Comment (carlocab) [inline Formula/llvm.rb:168]: Oops:
```
make: *** No rule to make target `install-xcode-toolchain/strip'.  Stop.
```
They have `/strip` targets for all their other `install` targets; I would've thought they'd have one for this too.

Comment (danielnachun) [reply Formula/llvm.rb:168]: Yeah that's a bummer.  At least we have the answer now.  Do you want me to wait to push to this again until you've had a chance to test your LLVM PR?

Comment (carlocab) [reply Formula/llvm.rb:168]: I'm probably going to sit on that for a while, but if could pull in the cosmetic changes from there to here (i.e. the ones that don't involve `C_INCLUDE_DIRS`) I'd appreciate it.

Comment (carlocab) [reply Formula/llvm.rb:168]: i.e. These:
```diff
diff --git a/Formula/llvm.rb b/Formula/llvm.rb
index 1bbe232ae6..a0bc619c42 100644
--- a/Formula/llvm.rb
+++ b/Formula/llvm.rb
@@ -76,7 +76,7 @@ class Llvm < Formula
 
     # we install the lldb Python module into libexec to prevent users from
     # accidentally importing it with a non-Homebrew Python or a Homebrew Python
-    # in a non-default prefix
+    # in a non-default prefix. See https://lldb.llvm.org/resources/caveats.html
     args = %W[
       -DLLVM_ENABLE_PROJECTS=#{projects.join(";")}
       -DLLVM_ENABLE_RUNTIMES=#{runtimes.join(";")}
@@ -99,6 +99,9 @@ class Llvm < Formula
       -DLLDB_PYTHON_RELATIVE_PATH=libexec/#{site_packages}
       -DLIBOMP_INSTALL_ALIASES=OFF
       -DCLANG_PYTHON_BINDINGS_VERSIONS=#{py_ver}
+      -DPACKAGE_VENDOR=#{tap.user}
+      -DPACKAGE_BUGREPORT=#{tap.issues_url}
+      -DCLANG_VENDOR_UTI=org.#{tap.user.downcase}.clang
     ]
 
     if MacOS.version >= :catalina
@@ -113,7 +116,7 @@ class Llvm < Formula
       args << "-DLLVM_BUILD_LLVM_C_DYLIB=ON"
       args << "-DLLVM_ENABLE_LIBCXX=ON"
       args << "-DLLVM_CREATE_XCODE_TOOLCHAIN=#{MacOS::Xcode.installed? ? "ON" : "OFF"}"
-      args << "-DRUNTIMES_CMAKE_ARGS=-DCMAKE_INSTALL_RPATH=@loader_path/../lib"
+      args << "-DRUNTIMES_CMAKE_ARGS=-DCMAKE_INSTALL_RPATH=#{rpath}"
 
       sdk = MacOS.sdk_path_if_needed
       args << "-DDEFAULT_SYSROOT=#{sdk}" if sdk
```

Comment (danielnachun) [reply Formula/llvm.rb:168]: Actually I see you marked that PR as a draft now.  I'll go ahead and push the fix for this, but feel free to cancel my run if it's causing problems.

Comment (carlocab) [reply Formula/llvm.rb:168]: Yup, I'll need to test that one locally a bit first. But if you could include some of the other changes (see above) here to save some CI runs, that'd be great.

Comment (danielnachun) [reply Formula/llvm.rb]: I think I got all the changes in now.  If you see anything wrong or missing I'll restart the job rather than wait.

**Comment by danielnachun** [comment]
We're all green here now!  When we've finalized this I will backport the relevant changes to `llvm@11`, and then work on the older versions after that.

**Review by carlocab — APPROVED** [review]
Merging these directly, as I don't think we need new bottles on macOS for the relevant changes.

**Comment by danielnachun** [comment]
Actually did we need to rebottle because we used `install/strip`?

**Comment by carlocab** [comment]
The new bottle is about 88MB smaller. I'm not sure it's worth carrying another ~1.5GB in (compressed) bottles for, but we can rebottle it if you feel strongly about it.

**Pull request merged**
